Flow startup errors -- better startup error handling with informative messages  (Part 1)

Vaadin 20.0.1:

Caused by: java.lang.NullPointerException
                at com.vaadin.flow.server.startup.LookupServletContainerInitializer.process(LookupServletContainerInitializer.java:64)
                at com.vaadin.flow.server.startup.ClassLoaderAwareServletContainerInitializer.lambda$onStartup$2(ClassLoaderAwareServletContainerIni
tializer.java:81)
                at com.vaadin.flow.server.startup.ClassLoaderAwareServletContainerInitializer.onStartup(ClassLoaderAwareServletContainerInitializer.java:122)
                at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5157)
                at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)



```
 @Override
    public void process(Set<Class<?>> classSet, ServletContext servletContext)
            throws ServletException {
        if (!classSet.contains(LookupInitializer.class)) {
            // this is a specific case for OSGi (PAX web): at some point it may
            // decide to apply ServletContainerInitializers for non WAR case
            // even though it has no any sense to do it because the {@code
            // classSet} is always empty. In normal environment if Servlet
            // container properly supports ServletContainerInitializers then
            // {@code classSet} has to contain at least LookupInitializer.
            // Otherwise LookupServletContainerInitializer should not be
            // executed at all.
            return;
        }
```


I am encountering this problem deploying to Tomcat8 an MPR project which quite obviously has a broken setup.

However. This implicit NPE should not happen.

Expected: I get some informative error message why Flow cannot start.

The code underneath also seems to handle NULL gracefully as if it was an empty set, it is the line
if (!classSet.contains(....   that is the problem here. And a missing informative error message.

--- a/flow-server/src/main/java/com/vaadin/flow/server/startup/LookupServletContainerInitializer.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/startup/LookupServletContainerInitializer.java
@@ -64,6 +64,15 @@ public class LookupServletContainerInitializer
     @Override
     public void process(Set<Class<?>> classSet, ServletContext servletContext)
             throws ServletException {
+        if (classSet == null) {
+            throw new ServletException(ServletContainerInitializer.class
+                    .getSimpleName() + " is called but the "
+                    + "provided set of classes is 'null'. "
+                    + LookupInitializer.class + " always presents "
+                    + "and has to be passed to the 'onStartup' method as an argument "
+                    + "in the set of classes if the servlet container supports Servlet 3.0 specification. "
+                    + "The propject configuration is broken somehow or you are using Servlet 3.0 incompatible container.");
+        }
         if (!classSet.contains(LookupInitializer.class)) {
             // this is a specific case for OSGi (PAX web): at some point it may
             // decide to apply ServletContainerInitializers for non WAR case

--- a/flow-server/src/test/java/com/vaadin/flow/server/startup/LookupServletContainerInitializerTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/startup/LookupServletContainerInitializerTest.java
@@ -213,6 +213,11 @@ public class LookupServletContainerInitializerTest {
         Assert.assertEquals(ArrayList.class, collection.iterator().next());
     }
 
+    @Test(expected = ServletException.class)
+    public void process_classSetIsNull_throws() throws ServletException {
+        initializer.process(null, Mockito.mock(ServletContext.class));
+    }
+
     private Lookup mockLookup(ServletContext context, Class<?>... classes)
             throws ServletException {
         ArgumentCaptor<Lookup> lookupCapture = ArgumentCaptor

